cou/steps/execute.py

```diff
@@ -89,25 +88,45 @@ async def _run_step(step: UpgradeStep, interactive: bool) -> None:
             await apply_plan(sub_step, interactive)
 
 
-async def apply_plan(plan: UpgradeStep, interactive: bool) -> None:
+async def apply_plan(plan: BaseStep, interactive: bool) -> None:
     """Apply the plan for upgrade.
 
     :param plan: Plan to be executed on steps.
-    :type plan: UpgradeStep
+    :type plan: BaseStep
     :param interactive:
     :type interactive: bool
     """
+    description = plan.description + " "
     result = ""
+    indicate_progress = True
+
+    # do nothing for empty upgrade step
+    if not plan:
+        return
+
+    # for UpgradePlan and its children class, do not show progress indicator
+    if isinstance(plan, UpgradePlan):
+        indicate_progress = False
+
+    # group and print all sub-steps with hierarchy for ApplicationUpgradePlan
+    if isinstance(plan, ApplicationUpgradePlan):
+        description = str(plan) + "\n"
+        if not interactive:
+            print(plan.description)
+
     while result.casefold() not in AVAILABLE_OPTIONS:
-        result = (await ainput(prompt(plan.description))).casefold() if interactive else "c"
+        if not interactive or plan.prompt is False:
```

Comment:
```suggestion
        if interactive is True or plan.prompt is False:
```

Reply:
Do you mean `if interactive is False ...`? If interactive is true we should prompt for input instead of setting result to `y`

Reply:
Sorry yes, I meant `interactive is False` only because you are doing `plan.prompt is False` and I prefer to stick with same usage. It means that you can use also `if not interactive or not plan.prompt`.